File: June/23-06-2024/Print Bracket Number.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++
class Solution {
  public:

    vector<int> bracketNumbers(string str) {
        // Your code Start
    vector<int> res(str.length(), 0); // Initialize the result vector
        stack<pair<int, int>> stk;
        int count = 0;
        
        for (int i = 0; i < str.length(); ++i) {
            if (str[i] == '(') {
                count++;
                stk.push({i, count});
                res[i] = count;
            } else if (str[i] == ')') {
                pair<int, int> p = stk.top();
                stk.pop();
                res[i] = p.second;
                res[p.first] = p.second;
            }
        }
        
        vector<int> bracketNum;
        for (int num : res) {
            if (num != 0) {
                bracketNum.push_back(num);
            }
        }
        
        return bracketNum;
      //Your Code End
    }
};

//{ Driver Code Starts.

int main() {

    ios_base::sync_with_stdio(0);
    cin.tie(NULL);
    cout.tie(NULL);

    int t;
    string tc;
    getline(cin, tc);
    t = stoi(tc);
    while (t--) {
        string s;
        getline(cin, s);

        Solution ob;

        vector<int> ans = ob.bracketNumbers(s);

        for (auto i : ans)
            cout << i << " ";

        cout << "\n";
    }

    return 0;
}
// } Driver Code Ends

File: June/23-06-2024/Print Bracket Number.js
//{ Driver Code Starts
// Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function printList(res, n) {
    let s = "";
    for (let i = 0; i < n; i++) {
        s += res[i];
        s += " ";
    }
    console.log(s);
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let s = readLine();

        let obj = new Solution();
        let res = obj.bracketNumbers(s);
        printList(res, res.length);
    }
} // } Driver Code Ends

// } Driver Code Ends


// User function Template for javascript

/**
 * @param {string} s
 * @returns {number[]}
 */

class Solution {
      bracketNumbers(str) {
        //Your Code Start
            let res = Array(str.length).fill(0);
        let stk = [];
        let count = 0;
        
        for (let i = 0; i < str.length; i++) {
            if (str[i] === '(') {
                count++;
                stk.push([i, count]);
                res[i] = count;
            } else if (str[i] === ')') {
                let [index, num] = stk.pop();
                res[i] = num;
                res[index] = num;
            }
        }
        
        let bracketNum = [];
        for (let num of res) {
            if (num !== 0) {
                bracketNum.push(num);
            }
        }
        
        return bracketNum;
        //Your Code End
    }
}
